### PR TITLE
C++ IR: Make instruction operand getters have only exact results

### PR DESCRIPTION
**Pull request by jbj**
This is my attempt to implement the changes we [discussed on Slack](https://semmle.slack.com/archives/C7NS4H8HY/p1559242384011500), making the default memory operand getters have a result only for definitions that _must exactly overlap_. This is a breaking API change, but I think the new behaviour is what most users were expecting in the first place.

I'd like to hear @dave-bartolomeo's initial comments before I refine this further.

To do:
- Bikeshedding predicate names. I've chosen `getDef` and `getAnyDef` because they are short and because the shortest of the two is the one most people should be using.
- Documenting, and ideally also testing, what it means for a definition to be non-exact. We need some examples in the documentation of `getDef` and `getAnyDef`.
- Considering whether some `getAnyDef` should become `getDef` in the SSA construction code.
- Benchmarking.
- Writing a change note.

### CODE REVIEW

**Comment by jbj** [comment]
I think `numberableInstruction` in `ValueNumbering.qll` needs to change so it only includes instructions whose operands are all congruent.

**Comment by jbj** [comment]
> I think `numberableInstruction` in `ValueNumbering.qll` needs to change so it only includes instructions whose operands are all congruent.

Or maybe not. Can the instructions listed in `numberableInstruction` have non-congruent parameters? Except for the case of https://github.com/jbj/ql/commit/d6a10ed81e7fc55c23eb2de439c0fc5c51a32f66, which I'll address in a separate PR.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.expected [hunk 1]
```diff
@@ -17,8 +17,6 @@
 | test.cpp:110:10:110:12 | Load: (reference dereference) | test.cpp:109:9:109:14 | Call: call to source |
 | test.cpp:126:8:126:19 | Convert: (const int *)... | test.cpp:120:9:120:20 | InitializeParameter: sourceArray1 |
 | test.cpp:126:8:126:19 | Load: sourceArray1 | test.cpp:120:9:120:20 | InitializeParameter: sourceArray1 |
-| test.cpp:132:22:132:23 | Load: m1 | test.cpp:122:18:122:30 | InitializeParameter: sourceStruct1 |
```
Comment (dave-bartolomeo) [inline cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.expected]: Were these false positives that were fixed, or false negatives that were introduced?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll [hunk 2]
```diff
@@ -1742,7 +1743,7 @@ class ChiInstruction extends Instruction {
    * Gets the operand that represents the new value written by the memory write.
    */
   final Instruction getPartial() {
-    result = getPartialOperand().getDefinitionInstruction()
+    result = getPartialOperand().getDef()
```
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll]: I just tested it on a comdb2 snapshot with this query:

```
import semmle.code.cpp.ir.IR

select
  strictcount(ChiInstruction i) as chiInstructions,
  count(ChiInstruction i | exists(i.getTotal())) as totals,
  count(ChiInstruction i | exists(i.getPartial())) as partials
```

The result: 174506, 174506, 174506. Does that mean there's a problem elsewhere with the overlap classification of chi instructions?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Operand.qll [hunk 3]
```diff
@@ -26,27 +26,63 @@ class Operand extends TOperand {
...
   }
 
+  /**
+   * Gets the `Instruction` whose result is the value of the operand. Unlike
+   * `getAnyDef`, this also has no result when `isDefinitionInexact` holds,
+   * which means that the resulting instruction must always be exactly the be
+   * the value of this operand.
+   */
+  final Instruction getDef() {
+    result = this.getAnyDef() and
```
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Operand.qll]: I tried this, and it just shuffled the DIL around a bit and didn't seem to make a difference. In both cases, `getDef` becomes a dispatch predicate and is never inlined even when the DIL clearly shows that there's an inlining opportunity like here:

```
Instruction::BinaryInstruction::getRight_dispred#2#ff(cached Instruction::BinaryInstruction this,
                                                      cached TInstruction#3 result) :-
  exists(Operand::RightOperand#2 receiver |
    Instruction::BinaryInstruction::getRightOperand_dispred#2#ff(this, receiver),
    Operand::Operand::getDef_dispred#2#ff(receiver, result)
  )
.
```

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll [hunk 4]
```diff
@@ -227,10 +230,11 @@ cached ValueNumber valueNumber(Instruction instr) {
 }
 
 /**
- * Gets the value number assigned to `instr`, if any. Returns at most one result.
+ * Gets the value number assigned to the exact definition of `op`, if any.
+ * Returns at most one result.
  */
 ValueNumber valueNumberOfOperand(Operand op) {
-  result = valueNumber(op.getDefinitionInstruction())
+  result = valueNumber(op.getDef())
```
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll]: It's used in `RangeAnalysis.qll`, and this PR introduces uses in `SignAnalysis.qll`.

File: cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeUtils.qll [hunk 5]
```diff
@@ -22,8 +22,8 @@ IntValue getConstantValue(Instruction instr) {
   result = getConstantValue(instr.(CopyInstruction).getSourceValue()) or
   exists(PhiInstruction phi |
     phi = instr and
-    result = max(PhiInputOperand operand | operand = phi.getAnOperand() | getConstantValue(operand.getDefinitionInstruction())) and
-    result = min(PhiInputOperand operand | operand = phi.getAnOperand() | getConstantValue(operand.getDefinitionInstruction()))
+    result = max(PhiInputOperand operand | operand = phi.getAnOperand() | getConstantValue(operand.getAnyDef())) and
```
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeUtils.qll]: That means we also need to require that the phi instruction has no non-congruent inputs. But this should be implicitly handled here thanks to the semantics of monotonic `max`: if any operand does not produce a value for `getConstantValue(operand.getDef())`, then the whole `max` will not have a value. So everything is good after your suggested change.

**Comment by jbj** [comment]
I think the phi handling in `ConstantAnalysis.qll` is wrong now. I'll investigate.

**Comment by jbj** [comment]
Added a commit to fix that.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll]: Oops, for some reason I thought the definition of the partial operand was always a virtual variable, when in fact it should always be the exact same location as the result of the store that caused the `Chi` to be inserted.

File: cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.expected (hunk 1 above)
Comment (jbj) [reply cpp/ql/test/library-tests/dataflow/dataflow-tests/test_ir.expected]: I consider them to be false positives. They were both flows from `x` to `x.a`. That should propagate taint but not data flow.

**Pull request merged**